src/server/pfs/server/api_server.go

```diff
@@ -384,97 +388,63 @@ type modifyFileSource interface {
+// SetCommit messages will result in an error.
+func (a *apiServer) modifyFile(ctx context.Context, uw *fileset.UnorderedWriter, server modifyFileSource) (int64, error) {
 	var bytesRead int64
 	for {
-		// TODO Validation.
-		if req != nil && req.Modification != nil {
-			switch mod := req.Modification.(type) {
-			case *pfs.ModifyFileRequest_PutFile:
-				var err error
-				var n int64
-				switch mod.PutFile.Source.(type) {
-				case *pfs.PutFile_RawFileSource:
-					n, err = putFileRaw(uw, server, mod.PutFile)
-				case *pfs.PutFile_TarFileSource:
-					n, err = putFileTar(uw, server, mod.PutFile)
-				case *pfs.PutFile_UrlFileSource:
-					n, err = putFileURL(ctx, uw, mod.PutFile)
-				}
-				if err != nil {
-					return bytesRead, err
-				}
-				bytesRead += n
-			case *pfs.ModifyFileRequest_DeleteFile:
-				if err := deleteFile(uw, mod.DeleteFile); err != nil {
-					return bytesRead, err
-				}
-			case *pfs.ModifyFileRequest_CopyFile:
-				cf := mod.CopyFile
-				if err := a.driver.copyFile(ctx, uw, cf.Dst, cf.Src, cf.Append, cf.Tag); err != nil {
-					return bytesRead, err
-				}
-			}
-		}
-		var err error
-		req, err = server.Recv()
+		msg, err := server.Recv()
 		if err != nil {
-			if errors.Is(err, io.EOF) {
-				return bytesRead, nil
+			if err == io.EOF {
+				break
 			}
 			return bytesRead, err
 		}
-	}
-}
-
-func putFileTar(uw *fileset.UnorderedWriter, server modifyFileSource, req *pfs.PutFile) (int64, error) {
-	src := req.Source.(*pfs.PutFile_TarFileSource).TarFileSource
-	tfsr := &tarFileSourceReader{
-		server: server,
-		r:      bytes.NewReader(src.Data),
-	}
-	tr := tar.NewReader(tfsr)
-	for {
-		hdr, err := tr.Next()
-		if err != nil {
-			if errors.Is(err, io.EOF) {
-				return tfsr.bytesRead, nil
+		switch mod := msg.Body.(type) {
+		case *pfs.ModifyFileRequest_AddFile:
+			var err error
+			var n int64
+			p := mod.AddFile.Path
+			t := mod.AddFile.Tag
+			switch src := mod.AddFile.Source.(type) {
+			case *pfs.AddFile_Raw:
+				n, err = putFileRaw(uw, p, t, src.Raw)
+			case *pfs.AddFile_Url:
+				n, err = putFileURL(ctx, uw, p, t, src.Url)
+			default:
+				// need to write empty data to path
+				n, err = putFileRaw(uw, p, t, &types.BytesValue{})
 			}
-			return tfsr.bytesRead, err
-		}
-		if hdr.Typeflag == tar.TypeDir {
-			continue
-		}
-		if err := uw.Put(hdr.Name, req.Tag, req.Append, tr); err != nil {
-			return tfsr.bytesRead, err
+			if err != nil {
+				return bytesRead, err
+			}
+			bytesRead += n
+		case *pfs.ModifyFileRequest_DeleteFile:
+			if err := deleteFile(uw, mod.DeleteFile); err != nil {
+				return bytesRead, err
+			}
+		case *pfs.ModifyFileRequest_CopyFile:
+			cf := mod.CopyFile
+			if err := a.driver.copyFile(ctx, uw, cf.Dst, cf.Src, cf.Append, cf.Tag); err != nil {
+				return bytesRead, err
+			}
+		case *pfs.ModifyFileRequest_SetCommit:
+			return bytesRead, errors.Errorf("cannot set commit")
```

Comment:
It might be clearer to change this to "cannot set commit more than once".

Reply:
It is also used in CreateFileset, which does not allow a commit to be set at any point.